### ISSUE TITLE
S922x安装jellyfin解码4K压力大

### ISSUE THREAD

**Issue by stevenliuit** [ophub/amlogic-s9xxx-armbian#326]
你好 我用jellyfin 解码4K视频 Cpu有一个飙升很高 其他平稳
但是在看视频的时候 4k会很费力 安卓系统的时候看不卡顿 ffmpeg也是最新的 
是不是mali的开源驱动解码有问题啊？

**Comment by ophub** [comment]
cpu占用多少

**Comment by stevenliuit** [comment]
![setcode](https://user-images.githubusercontent.com/4044994/169774247-83aa8467-f04b-416c-a44c-3b288e3164a2.png)
![codepath](https://user-images.githubusercontent.com/4044994/169774246-8941af97-6aa7-4e00-8ee0-970929498cae.png)
![htop](https://user-images.githubusercontent.com/4044994/169774259-7e42b0d8-d3fb-4ffa-bd3d-1cecade32a84.png)
![htop2](https://user-images.githubusercontent.com/4044994/169774270-ac81d1a5-cc2f-43fb-987d-370818bf75da.png)
![htop3](https://user-images.githubusercontent.com/4044994/169774282-577630fa-76df-43ef-b850-cc159881153f.png)
上面是gpu访问的/dev/dri/renderD128解码
和jellyfin配置

**Comment by stevenliuit** [comment]
应该是对的 我的是gtking 就直接选的你那个gtking的内核 cpu是s922x wifi5版本的

**Comment by stevenliuit** [comment]
Linux armbian 5.15.36-flippy-72+o #112 SMP PREEMPT Wed Apr 27 21:29:33 CST 2022 aarch64 aarch64 aarch64 GNU/Linux

**Comment by ophub** [comment]
xxx这个是要换成你要查找的模块

你换成5.17的内核试试，在5.15内核下你的内存识别的少了

**Comment by stevenliuit** [comment]
# CONFIG_I2C_NVIDIA_GPU is not set
CONFIG_VGA_ARB_MAX_GPUS=16
# CONFIG_DRM_AMDGPU is not set
CONFIG_DRM_VIRTIO_GPU=m
gpu的查看到了这几项

**Comment by stevenliuit** [comment]
> xxx这个是要换成你要查找的模块
> 
> 你换成5.17的内核试试，在5.15内核下你的内存识别的少了

我试试5.17的内核 thx

**Comment by stevenliuit** [comment]
下载列表里只有
[Armbian_22.05.0-trunk_Odroidn2_jammy_edge_5.17.5.img.gz](https://github.com/ophub/amlogic-s9xxx-armbian/releases/download/Armbian_Aml_jammy_server_05.21.1222/Armbian_22.05.0-trunk_Odroidn2_jammy_edge_5.17.5.img.gz)
这个是5.17的 我的那个型号的没有啊 只有5.15和5.10

**Comment by ophub** [comment]
`armbian-update 5.17.9`

各系列的内核都可以直接更新切换，不用重新刷固件。

**Comment by stevenliuit** [comment]
> `armbian-update 5.17.9`
> 
> 各系列的内核都可以直接更新切换，不用重新刷固件。

嗯嗯 我试试

**Comment by ophub** [comment]
我搜了半天Beelink GT King Pro的评测文章和视频，都是一顿夸奖4K流畅播放的，但是没找到他们评测跑4K时的CPU占有率。你可以找找这方面的文章，看看你的CPU占有率是否符合情况。

**Comment by stevenliuit** [comment]
> 

我怀疑是gpu在linux里的表现不行 好像是gpu开源的不如闭源的给力 我找找文献

**Comment by stevenliuit** [comment]
网站找到了一个说驱动的


本人安装了Armbian_19.11.3_Aml-g12_stretch_legacy_5.3.0_20191126的系统后 装OMV ,折腾jellyfin后想硬件转码,不过没有折腾出来查下相关资料linux内核 5.4  https://git.kernel.org/pub/scm/l ... son/vdec?h=v5.4 这部分硬解的代码还是staging中 核心中代码也只是mpeg1,2的 没有264 265的,通过amlogic官网的 找到这里关于硬解的介绍What about multimedia support ?Here is current status on Multimedia support :GPU : Is supported on S905, S905X/W/L, S905D, S805X and S912 with Lima and PanfrostLinux drivers upstream since v5.2Mesa support in master : https://gitlab.freedesktop.org/mesa/mesaClosed Source mali driver using https://github.com/superna9999/meson_gx_mali_450 driver and https://github.com/superna9999/amlogic-meson-mali binaires (The ARM provided HiKey binaries are also functional)HDMI Output up to 4k is working on S905, S905X/W/L, S905D, S912, S805X, S905X2, S905Y2, S905D2, S922X, A311D, S905X3, S905D3, S905Y3 (Will need WiP patches for 4k420 support)Video Decoding is being finalized, initial V4L2 driver for MPEG2 decoding has been pushed and is ready to be merged when the firmwares (https://github.com/LibreELEC/meson-firmware) licence is figured out.All of this is being integrated in a specific LibreELEC branch at : https://github.com/LibreELEC/LibreELEC.tv/tree/amlogic and https://github.com/chewitt/LibreELEC.tv/tree/amlogic
--


本人安装了Armbian_19.11.3_Aml-g12_stretch_legacy_5.3.0_20191126的系统后 装OMV ,折腾jellyfin后想硬件转码,不过没有折腾出来查下相关资料

linux内核 5.4  
[https://git.kernel.org/pub/scm/l ... son/vdec?h=v5.4](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/staging/media/meson/vdec?h=v5.3-rc5) 这部分[硬解](http://www.linux-meson.com/doku.php)的代码还是staging中 核心中代码也只是mpeg1,2的 没有264 265的,通过amlogic官网的 找到这里关于硬解的介绍

What about multimedia support ?
Here is current status on Multimedia support :

GPU : Is supported on S905, S905X/W/L, S905D, S805X and S912 with Lima and Panfrost
Linux drivers upstream since v5.2
Mesa support in master : https://gitlab.freedesktop.org/mesa/mesa
Closed Source mali driver using https://github.com/superna9999/meson_gx_mali_450 driver and https://github.com/superna9999/amlogic-meson-mali binaires (The ARM provided HiKey binaries are also functional)
HDMI Output up to 4k is working on S905, S905X/W/L, S905D, S912, S805X, S905X2, S905Y2, S905D2, S922X, A311D, S905X3, S905D3, S905Y3 (Will need WiP patches for 4k420 support)
Video Decoding is being finalized, initial V4L2 driver for MPEG2 decoding has been pushed and is ready to be merged when the firmwares (https://github.com/LibreELEC/meson-firmware) licence is figured out.
All of this is being integrated in a specific LibreELEC branch at : https://github.com/LibreELEC/LibreELEC.tv/tree/amlogic and https://github.com/chewitt/LibreELEC.tv/tree/amlogic

地址：https://www.right.com.cn/forum/forum.php?mod=viewthread&tid=1723601&mobile=no

**Comment by ophub** [comment]
我问问f大吧，对内核的配置我不熟悉，我问问他有什么模块可以给硬解码提供服务。

**Comment by stevenliuit** [comment]
用命令升级了armbian-update 5.17.5
升级完 开机黑屏了

**Comment by ophub** [comment]
拔电重插

**Comment by stevenliuit** [comment]
ip获取不到 hdmi没显示 感觉是废了 电拔了重插的 也没跑码了

**Comment by ophub** [comment]
5.17.9，不是5，你刚才输入显示下载安装成功了？

**Comment by stevenliuit** [comment]
说错了 是这个
root@armbian:~# armbian-update 5.17.9
Ready to update, please wait...
custom_kernel_name: 5.17.9-flippy-73+
K510 [ 1 ]
Unpack [ 5.17.9-flippy-73+ ] related files ...
02.01 Unpack [ boot-5.17.9-flippy-73+.tar.gz ] complete.
02.02 Unpack [ dtb-amlogic-5.17.9-flippy-73+.tar.gz ] complete.
02.03 Unpack [ modules-5.17.9-flippy-73+.tar.gz ] complete.

gzip: stdin: unexpected end of file
tar: Child returned status 1
tar: Error is not recoverable: exiting now
02.04 Unpack [ header-5.17.9-flippy-73+.tar.gz ] complete.
The u-boot-gtkingpro-rev-a.bin file copy is complete.
Successfully updated, automatic restarting...

**Comment by ophub** [comment]
我的神，这都报错了，没安装成功

**Comment by stevenliuit** [comment]
> 5.17.9，不是5，你刚才输入显示下载安装成功了？

是5.17.9 邪门了 不过我看了 
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/staging/media/meson/vdec?h=v5.18
5.10到5.18 直邮h264解码 没有h265好像

**Comment by stevenliuit** [comment]
> Error is not recoverable: exiting now

发现了 哈哈  Error is not recoverable: exiting now 这句话 我滴神 我重新刷系统吧

**Comment by ophub** [comment]
![Snip20220523_3](https://user-images.githubusercontent.com/68696949/169793929-a6b76a49-d871-4e69-80f9-8d5b735bc429.png)


我刚在线更新了5.17.9，内核包是完整的，你本地没有安装解压缩软件包？

**Comment by ophub** [comment]
sudo apt-get install -y tar

**Comment by ophub** [comment]
看来我得升级一下更新内核的这个脚本，先检查本地是否安装了解压缩软件包。

**Comment by stevenliuit** [comment]
> 

刚才我试了 能升级 有一个升级成功了 默认升级到了5.15.41是成功的 指定后反而失败了 我先按你的试试 我先刷机 镜像下载好了 这就刷

**Comment by stevenliuit** [comment]
> 看来我得升级一下更新内核的这个脚本，先检查本地是否安装了解压缩软件包。

应该是在线包下载报错导致的 第一次执行报错了

**Comment by stevenliuit** [comment]
第一次执行报错这个了
root@armbian:~# armbian-update 5.17.9
Ready to update, please wait...
Kernel version: 5.17.9
Start downloading kernel to [ /root ]...
01.01 The boot file download complete.
01.02 The dtb file download complete.
01.03 The modules file download complete.
 [Error] 01.04 The header file failed to download.
然后我就又执行了一次 就废了

**Comment by ophub** [comment]
脚本不够严谨，得多做几个判断，以后加个md5值比较，以免出现你这样文件下载不完整还继续安装造成的损失。

**Comment by stevenliuit** [comment]
> 
哈哈 没事 生命在于折腾
meson-g12b-gsking-x.dtb和meson-g12b-gtking-oc.dtb 
这个oc和x有啥区别啊

**Comment by ophub** [comment]
oc就是超频的意思，你选择这个试试，超频性能更强一些

**Comment by stevenliuit** [comment]
> 

ok 我试试 改完这就尝试 不成功则成仁 哈哈

**Comment by ophub** [comment]
稍等

**Comment by stevenliuit** [comment]
![image](https://user-images.githubusercontent.com/4044994/169797978-4e8b9dc2-f2ef-49d8-92fe-55cb885a29aa.png)
这几个我应该下载哪个啊 我上次下载的忘了
我看有reva 我记得不是这个就是jammy

**Comment by stevenliuit** [comment]
> 你在usb里面改/boot/uEnv.txt，改那个dtb名称，改完重启可以正常使用再写入

嗯嗯 我试试

**Comment by ophub** [comment]
Armbian_22.05.0_Aml_s922x_jammy_5.10.117_server_2022.05.21.img.gz

**Comment by stevenliuit** [comment]
> Armbian_22.05.0_Aml_s922x_jammy_5.10.117_server_2022.05.21.img.gz

嗯嗯 我试试 先吃饭 吃完了继续鼓捣

**Comment by ophub** [comment]
硬解模块是 meson_vdec， 可用modprobe meson_vdec装载，但还是需要应用软件支持才行。

**Comment by stevenliuit** [comment]
> 我问了一下f大，他说硬解模块是 meson_vdec， 可用`modprobe meson_vdec`装载，但还是需要应用软件支持才行。你看看jellyfin是否支持

jellyfin支持的硬解有这几个 涉及到我知识盲区了
AMD AMF
Intel Quick Sync
MediaCodec Android
OpenMAX OMX
Nvidia NVENC
Video Acceleration API (VAAPI)
Exynos V4L2 MFC
Video ToolBox
这几项 

其中网上用的最多的是
Video Acceleration API (VAAPI)
我看咱们这个固件有
/dev/dri/renderD128
大多数是调用这个硬解的 其他的还真不了解 我装完了 用modprobe meson_vdec试试

**Comment by stevenliuit** [comment]
找到了一篇 大概意思是没支持
https://forum.odroid.com/viewtopic.php?p=343973

**Comment by stevenliuit** [comment]
内核升级到了5.17.9
root@armbian:/opt# uname -a
Linux armbian 5.17.9-flippy-73+ #47 SMP PREEMPT Wed May 18 17:31:09 CST 2022 aarch64 aarch64 aarch64 GNU/Linux

ffmpeg升级到了5.0.1最新版
root@armbian:/opt# /usr/lib/jellyfin-ffmpeg/ffmpeg -version
ffmpeg version 5.0.1-Jellyfin Copyright (c) 2000-2022 the FFmpeg developers
built with gcc 11 (Ubuntu 11.2.0-17ubuntu1)
configuration: --prefix=/usr/lib/jellyfin-ffmpeg --target-os=linux --extra-version=Jellyfin --disable-doc --disable-ffplay --disable-ptx-compression --disable-shared --disable-libxcb --disable-sdl2 --disable-xlib --enable-lto --enable-gpl --enable-version3 --enable-static --enable-gmp --enable-gnutls --enable-libdrm --enable-libass --enable-libfreetype --enable-libfribidi --enable-libfontconfig --enable-libbluray --enable-libmp3lame --enable-libopus --enable-libtheora --enable-libvorbis --enable-libdav1d --enable-libwebp --enable-libvpx --enable-libx264 --enable-libx265 --enable-libzvbi --enable-libzimg --enable-libfdk-aac --toolchain=hardened --enable-cross-compile --arch=arm64 --cross-prefix=/usr/bin/aarch64-linux-gnu-
libavutil      57. 17.100 / 57. 17.100
libavcodec     59. 18.100 / 59. 18.100
libavformat    59. 16.100 / 59. 16.100
libavdevice    59.  4.100 / 59.  4.100
libavfilter     8. 24.100 /  8. 24.100
libswscale      6.  4.100 /  6.  4.100
libswresample   4.  3.100 /  4.  3.100
libpostproc    56.  3.100 / 56.  3.100

尝试安装OpenCV不成功
测试解码 cpu还是标高 用的OC的dtb文件
![image](https://user-images.githubusercontent.com/4044994/169852806-3b7564b3-b477-4a36-b2f1-076758c7b203.png)

**Comment by JFLim1** [comment]
> I see that your picture shows that the memory is 3.26G, which is not correctly identified.

Hi @ophub 

I have highlighted before that when your kernel-5.15.y-flippy-xx+o is being booted **without** using u-boot.ext the RAM indicated is always 3.26GB instead of 3.70GB on GT King Pro.  

If you boot 5.15.y-flippy **using u-boot.ext** htop will indicate RAM is 3.70GB.

**Comment by ophub** [comment]
> > I see that your picture shows that the memory is 3.26G, which is not correctly identified.
> 
> Hi @ophub
> 
> I have highlighted before that when your kernel-5.15.y-flippy-xx+o is being booted **without** using u-boot.ext the RAM indicated is always 3.26GB instead of 3.70GB on GT King Pro.
> 
> If you boot 5.15.y-flippy **using u-boot.ext** htop will indicate RAM is 3.70GB.

There are differences according to different devices. I didn't add it when I used it, and the memory recognition is normal.

![Snip20220524_4](https://user-images.githubusercontent.com/68696949/169982876-6b43fb92-8f9e-400e-b424-bc87841b7fa6.png)

**Comment by JFLim1** [comment]
> There are differences according to different devices. I didn't add it when I used it, and the memory recognition is normal.

Interesting.  How do I recover the loss RAM? Device is Beelink GT King Pro (Rev A). Armbian-Jammy-5.15.41-flippy-73+o


```
jfl@armbian:~$ uname -a
Linux armbian 5.15.41-flippy-73+o #123 SMP PREEMPT Wed May 18 17:08:50 CST 2022 aarch64 aarch64 aarch64 GNU/Linux
```

```
jfl@armbian:~$ cat /etc/ophub-release
VERSION_CODEID='ubuntu'
VERSION_CODENAME='jammy'
FDTFILE='meson-g12b-gtking-pro.dtb'
UBOOT_OVERLOAD='u-boot-gtkingpro-rev-a.bin'
MAINLINE_UBOOT='/usr/lib/u-boot/'
ANDROID_UBOOT='/usr/lib/u-boot/'
SOC='s922x-reva'
PACKAGED_DATE='2022-04-30'
KERNEL_VERSION='5.15.41'
K510='0'
```

```
jfl@armbian:~$ ls /boot
aml_autoscript                  u-boot-n1.bin
aml_autoscript.cmd              u-boot-odroid-n2.bin
Android                         u-boot-p201.bin
armbian_first_run.txt.template  u-boot-p212.bin
boot.bmp                        u-boot-s905.bin
boot.cmd                        u-boot-s905x2-s922.bin
boot-emmc.cmd                   u-boot-s905x-s912.bin
boot-emmc.ini                   u-boot.sd
boot-emmc.scr                   u-boot-sei510.bin
boot.ini                        u-boot-sei610.bin
boot.scr                        u-boot-tx3-bz.bin
config-5.15.41-flippy-73+o      u-boot-tx3-qz.bin
dtb                             u-boot-u200.bin
emmc_autoscript                 u-boot-ugoos-x3.bin
emmc_autoscript.cmd             u-boot.usb
initrd.img-5.15.41-flippy-73+o  u-boot-x96max.bin
LOST.DIR                        u-boot-x96maxplus.bin
s905_autoscript                 u-boot-zyxq.bin
s905_autoscript.cmd             uEnv.txt
System.map-5.15.41-flippy-73+o  uInitrd
u-boot-e900v22c.bin             uInitrd-5.15.41-flippy-73+o
u-boot-gtking.bin               vmlinuz-5.15.41-flippy-73+o
u-boot-gtkingpro.bin            zImage
u-boot-gtkingpro-rev-a.bin
```


![image](https://user-images.githubusercontent.com/75384411/169986475-b17d8c1f-6192-4fe7-b6d6-7c24b992ae1e.png)

**Comment by ophub** [comment]
![Snip20220524_5](https://user-images.githubusercontent.com/68696949/169990295-5488614b-9ab4-49fd-89c7-e165f5b6a189.png)

**Comment by JFLim1** [comment]
> This device has never found a user, and I have not tested whether this mainline u-boot can be written to emmc for use. If you know how to restore the system, `you can test if it works.`

For the moment this device is also being used as home entertainment for family using Android to stream video and as a daily mini pc (linux). so it is not convenient for me to install mailine u-boot on the EMMC.

But if it works by install the mainline u-boot on USB Flash boot image then I would like to test it out with your guidance on how to install the mainline u-boot on the USB Flash drive.  Thanks.

**Comment by stevenliuit** [comment]
> > This device has never found a user, and I have not tested whether this mainline u-boot can be written to emmc for use. If you know how to restore the system, `you can test if it works.`
> 
> For the moment this device is also being used as home entertainment for family using Android to stream video and as a daily mini pc (linux). so it is not convenient for me to install mailine u-boot on the EMMC.
> 
> But if it works by install the mainline u-boot on USB Flash boot image then I would like to test it out with your guidance on how to install the mainline u-boot on the USB Flash drive. Thanks.

If you want to restore the system by writing emmc, download Beelink_GT-King_Pro_911P0_090720_Amlo.zip from the official website, this kind of similar flashing package, and then use Amlogic USB Burning Tool to restore the Android system.
Or check out other people's videos using: https://www.youtube.com/watch?v=LuCTH0NRi3Y

**Comment by ophub** [comment]
如果不支持就先这样吧

**Comment by JFLim1** [comment]
> Download the latest version of the kernel update script
> wget https://raw.githubusercontent.com/ophub/amlogic-s9xxx-armbian/main/build-armbian/common-files/rootfs/usr/sbin/armbian-update -O /usr/sbin/armbian-update

@ophub ,

Noted. Have updated armbian-update with the above command.



> # Add mainline uboot
> vi /etc/ophub-release
> MAINLINE_UBOOT='/usr/lib/u-boot/gtkingpro-u-boot.bin.sd.bin'

Just to clarify, if I add the above "MAINLINE_UBOOT='/usr/lib/u-boot/gtkingpro-u-boot.bin.sd.bin'" to /etc/ophub-release and issue the command "sudo armbian-update 5.15.43"

Will it automatically install the mainline u-boot into the EMMC?  Will the Android in EMMC also be wipeout or Android can still boot?

**Comment by ophub** [comment]
Installed on the device running on your current system. Don't try it if you're on USB, it only makes sense on EMMC.

**Comment by JFLim1** [comment]
> Installed on the device running on your current system. Don't try it if you're on USB, it only makes sense on EMMC.
> 
> If you want to test whether the emmc supports writing to the mainline u-boot, you can use armbian-ddbr to back up the Android system in the current device emmc to the usb, and you can use this backup to restore the Android system later.

@ophub,
Thanks.  For the time being will stick with booting from USB.  The device is being used daily for home entertainment -- Android.

**Comment by ophub** [comment]
https://github.com/ophub/amlogic-s9xxx-armbian/tree/main/build-armbian/armbian-docs#113-how-to-restore-the-original-android-tv-system

I just added the instructions for armbian-ddbr for you

**Comment by JFLim1** [comment]
> I just added the instructions for armbian-ddbr for you

Thank you.  You are doing a fantastic job in supporting the Amlogic TV Box community. 

Device is being use daily for streaming video on Android by other members so it is not convenient currently to erase Android eventhough it looks relatively safe and easy to restore Android on emmc.

So I take it currently with kernel 5.15-flippy (boot **without** u-boot.ext) on GT King Pro Rev A the RAM will available will be 3.26GB without mainline u-boot.on EMMC.

To answer your question whether GT King Pro Rev A can install mainline u-boot on EMMC, if not mistaken spikerguy from Manjaro Team also has a GT King Pro Rev A and he has install mainline Manjaor Arm on EMMC.
